sway/input/text_input.c

```diff
@@ -255,6 +267,240 @@ static void relay_handle_text_input(struct wl_listener *listener,
+		// For now we'll only check the overlay and top layer
+		struct wlr_scene_node *node;
+		wl_list_for_each (node, &output->layers.shell_top->children, link) {
+			struct sway_layer_surface *surface = scene_descriptor_try_get(node,
+				SWAY_SCENE_DESC_LAYER_SHELL);
+			if (!surface) {
+				continue;
+			}
+			if (surface->layer_surface == layer) {
+				return surface;
+			}
+		}
+		wl_list_for_each (node, &output->layers.shell_overlay->children, link) {
+			struct sway_layer_surface *surface = scene_descriptor_try_get(node,
+				SWAY_SCENE_DESC_LAYER_SHELL);
+			if (!surface) {
+				continue;
+			}
+			if (surface->layer_surface == layer) {
+				return surface;
+			}
+		}
+	}
+	return NULL;
+}
+
+
+static void input_popup_update(struct sway_input_popup *popup) {
+	struct sway_text_input *text_input =
+		relay_get_focused_text_input(popup->relay);
+
+	if (text_input == NULL || text_input->input->focused_surface == NULL) {
+		return;
+	}
+
+	if (!popup->popup_surface->surface->mapped) {
+		return;
+	}
+
+	if (popup->scene_tree != NULL) {
+		wlr_scene_node_destroy(&popup->scene_tree->node);
+		if (popup->desc.relative != NULL) {
+			wlr_scene_node_destroy(popup->desc.relative);
+		}
+
+	}
+
+	bool cursor_rect = text_input->input->current.features
+		& WLR_TEXT_INPUT_V3_FEATURE_CURSOR_RECTANGLE;
+	struct wlr_surface *focused_surface = text_input->input->focused_surface;
+	struct wlr_box cursor_area = text_input->input->current.cursor_rectangle;
+
+	struct wlr_box output_box;
+	struct wlr_box parent;
+	struct wlr_layer_surface_v1 *layer_surface =
+		wlr_layer_surface_v1_try_from_wlr_surface(focused_surface);
+	struct wlr_scene_tree *relative;
+	if (layer_surface != NULL) {
+		struct sway_layer_surface *layer =
+			find_layer_by_surface(layer_surface);
+		if (layer == NULL) {
+			return;
+		}
+
+		struct wlr_output* output = layer->layer_surface->output;
+		wlr_output_layout_get_box(root->output_layout, output, &output_box);
+		int lx, ly;
+		wlr_scene_node_coords(&layer->tree->node, &lx, &ly);
+		parent.x = lx;
+		parent.y = ly;
+		popup->scene_tree = wlr_scene_subsurface_tree_create(root->layers.popup, popup->popup_surface->surface);
+		popup->desc.view = NULL;
+		relative = wlr_scene_tree_create(layer->scene->tree);
+	} else {
+		struct sway_view *view = view_from_wlr_surface(focused_surface);
+		int lx, ly;
+		wlr_scene_node_coords(&view->scene_tree->node, &lx, &ly);
+		struct wlr_output *output = wlr_output_layout_output_at(root->output_layout,
+			view->container->pending.content_x + view->geometry.x,
+			view->container->pending.content_y + view->geometry.y);
+		wlr_output_layout_get_box(root->output_layout, output, &output_box);
+		parent.x = lx;
+		parent.y = ly;
+
+		parent.width = view->geometry.width;
+		parent.height = view->geometry.height;
+		popup->scene_tree = wlr_scene_subsurface_tree_create(root->layers.popup, popup->popup_surface->surface);
+		popup->desc.view = view;
+		relative = wlr_scene_tree_create(view->scene_tree);
```

Comment:
The assignment here is the same for both branches, I would factor that out move it down (or up, because you also aren't checking for allocation failure here.)

Reply:
This still needs to be addressed. I think what @Nefsen402 suggests is something among these lines:

```c
	struct wlr_scene_tree *relative_parent;
	if (layer_surface != NULL) {
		struct sway_layer_surface *layer = layer_surface->data;
		if (layer == NULL) {
			return;
		}

		relative_parent = layer->scene->tree;

		struct wlr_output* output = layer->layer_surface->output;
		wlr_output_layout_get_box(root->output_layout, output, &output_box);
		int lx, ly;
		wlr_scene_node_coords(&layer->tree->node, &lx, &ly);
		parent.x = lx;
		parent.y = ly;
	} else {
		struct sway_view *view = view_from_wlr_surface(focused_surface);
		relative_parent = view->scene_tree;

		int lx, ly;
		wlr_scene_node_coords(&view->scene_tree->node, &lx, &ly);
		struct wlr_output *output = wlr_output_layout_output_at(root->output_layout,
			view->container->pending.content_x + view->geometry.x,
			view->container->pending.content_y + view->geometry.y);
		wlr_output_layout_get_box(root->output_layout, output, &output_box);
		parent.x = lx;
		parent.y = ly;
		parent.width = view->geometry.width;
		parent.height = view->geometry.height;
		popup->desc.view = view;
	}

	struct wlr_scene_tree *relative = wlr_scene_tree_create(relative_parent);
	popup->desc.relative = &relative->node;

	popup->scene_tree = wlr_scene_subsurface_tree_create(root->layers.popup, popup->popup_surface->surface);
```